Org habit stats

ml729 opened the pull request:
### Brief summary of what the package does

View statistics, a calendar, and bar graphs of your org-habits in Emacs org-mode.

### Direct link to the package repository

https://github.com/ml729/org-habit-stats

### Your association with the package

Author, maintainer.

### Relevant communications with the upstream package maintainer

None needed.

### Checklist

<!-- Please confirm by replacing `[]` with `[x]`: -->

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

<!-- After submitting, please fix any problems the CI reports. -->

riscy commented:
Thank you for this!  Very quick first pass --

`org-habit-stats.el` with byte-compile using Emacs 28.1:
```
org-habit-stats.el:448:62: Warning: Unused lexical argument `history'
org-habit-stats.el:448:62: Warning: Unused lexical argument `habit-data'
In org-habit-stats-unstreak:
org-habit-stats.el:473:64: Warning: Unused lexical argument `history'
org-habit-stats.el:473:64: Warning: Unused lexical argument `habit-data'
In org-habit-stats--record-streak-full:
org-habit-stats.el:481:75: Warning: Unused lexical argument `history-rev'
org-habit-stats.el:481:75: Warning: Unused lexical argument `habit-data'
In org-habit-stats-record-streak-format:
org-habit-stats.el:515:66: Warning: Unused lexical argument `history'
In org-habit-stats--N-day-percentage:
org-habit-stats.el:561:74: Warning: Unused lexical argument `habit-data'
In org-habit-stats-alltime-percentage:
org-habit-stats.el:583:64: Warning: Unused lexical variable `repeat-len'
In org-habit-stats-30-day-total:
org-habit-stats.el:595:58: Warning: Unused lexical argument `habit-data'
In org-habit-stats-365-day-total:
org-habit-stats.el:602:59: Warning: Unused lexical argument `habit-data'
In org-habit-stats-alltime-total:
org-habit-stats.el:609:59: Warning: Unused lexical argument `history'
org-habit-stats.el:609:59: Warning: Unused lexical argument `history-rev'
In org-habit-stats--exp-smoothing-list-full:
org-habit-stats.el:616:70: Warning: Unused lexical argument `history-rev'
org-habit-stats.el:616:70: Warning: Unused lexical argument `habit-data'
In org-habit-stats-make-calendar-buffer:
org-habit-stats.el:800:41: Warning: Unused lexical variable `completed-dates'
In org-habit-stats-graph-completions-per-month:
org-habit-stats.el:928:73: Warning: Unused lexical argument `history-rev'
org-habit-stats.el:928:73: Warning: Unused lexical argument `habit-data'
In org-habit-stats-graph-completions-per-week:
org-habit-stats.el:942:72: Warning: Unused lexical argument `history-rev'
org-habit-stats.el:942:72: Warning: Unused lexical argument `habit-data'
In org-habit-stats-graph-completions-per-weekday:
org-habit-stats.el:963:75: Warning: Unused lexical argument `history-rev'
org-habit-stats.el:963:75: Warning: Unused lexical argument `habit-data'
In org-habit-stats-insert-calendar:
org-habit-stats.el:1412:41: Warning: Unused lexical argument `habit-data'
In org-habit-stats-create-habit-buffer:
org-habit-stats.el:1501:85: Warning: Unused lexical variable `buff'
```

`org-habit-stats.el` with package-lint 20221003.1636:
```
1 issue found:
199:60: warning: Use variable `user-emacs-directory' or function `locate-user-emacs-file' instead of a literal path to the Emacs user directory or files.
```

`org-habit-stats.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- org-habit-stats.el#L659: There is no need to quote lambdas (neither #' nor ')
- org-habit-stats.el#L679: It's safer to sharp-quote function names; use `#'`
- org-habit-stats.el#L679: It's safer to sharp-quote function names; use `#'`
- org-habit-stats.el#L1557: It's safer to sharp-quote function names; use `#'`
- org-habit-stats.el#L1047: It's safer to sharp-quote function names; use `#'`
- org-habit-stats.el#L1055: It's safer to sharp-quote function names; use `#'`
- org-habit-stats.el#L1558: It's safer to sharp-quote function names; use `#'`
- org-habit-stats.el#L1059: It's safer to sharp-quote function names; use `#'`
- org-habit-stats.el#L1754: Ensure this `fset` isn't being used as a surrogate `defalias`
- org-habit-stats.el#L1587: Could this point/point-max comparison use `eobp`?
- org-habit-stats.el#L1656: Could this point/point-max comparison use `eobp`?
- org-habit-stats.el#L1605: Could this point/point-min comparison use `bobp`?
- org-habit-stats.el#L1675: Could this point/point-min comparison use `bobp`?
- org-habit-stats.el#L1557: Loading a package should rarely add hooks
- org-habit-stats.el#L1047: Loading a package should rarely add advice
- org-habit-stats.el#L1558: Loading a package should rarely add advice
- org-habit-stats.el#L442: No `format` required; `error` takes an f-string
```

riscy commented on recipes/org-habit-stats:
```diff
@@ -0,0 +1,3 @@
+(org-habit-stats :repo "ml729/org-habit-stats"
+                 :fetcher github
+                 :files ("org-habit-stats.el"))
```
Also I believe this recipe can be simplified to the default (giving you some wiggle room in the future):
```suggestion
(org-habit-stats :repo "ml729/org-habit-stats" :fetcher github)
```

ml729 commented:
Thanks for the feedback! I have made most of the corrections.

For the package-lint issue, it's because I wrote ".emacs.d" in a string (there's a feature that displays rewards if you reach a certain habit streak, one of the example rewards I put as kind of a joke is "Configure my .emacs.d") It does not affect any functionality.

As for the advices I added, one of them is to fix a bug in chart.el function 'chart-translate-xpos. This should probably be fixed in chart.el, but since chart.el is built into emacs, older Emacs versions will still have the bug regardless. I hope it's okay if I leave the fix here.
The other advice I added is to change the behavior of 'chart-draw-line when it is used inside 'chart-draw-axis (to use a custom character to draw the axes), and the advice is removed after the function is applied.

I'm using fset on a variable (not on the symbol explicitely) which I think means it is not a surrogate defalias?

riscy commented:
> one of them is to fix a bug in chart.el function 'chart-translate-xpos. This should probably be fixed in chart.el, but since chart.el is built into emacs, older Emacs versions will still have the bug regardless. I hope it's okay if I leave the fix here.

The risk is that future versions of Emacs may bring in different expectations/semantics to `chart-translate-xpos` which could be incompatible with your override (q.v. Hyrum's law, I guess).  I think it would be better for users if you just wrote your own function with the behavior your desire, instead of overriding a builtin.

> The other advice I added is to change the behavior of 'chart-draw-line when it is used inside 'chart-draw-axis (to use a custom character to draw the axes), and the advice is removed after the function is applied.

I think this is better than globally overriding the builtin function, especially since you've taken the care to `unwind-protect`, but it seems fragile.  I'd recommend skipping the advice and just writing your own modified copies of these functions.  There'll be a smaller blast radius if something goes wrong.

The .emacs.d and `fset` look totally fine to me -- I should have omitted those but had my foot on the gas. :)

ml729 commented:
Gotcha, I removed the advices and rewrote the relevant functions.

riscy commented:
Thank you!  And welcome to MELPA. :)
